MDM docs: Add macOS setup page

noahtalerman opened the pull request:
- Add "macOS setup" page. This naming matches the naming we'll use in the UI
- Add "Bootstrap package" section 
  - "signing the package" step needs better instructions (how do I generate a certificate? Once I have a cert, how do I sign the package?). These will be addressed in a later pass
  - fleetctl CLI for bootstrap package was released in 4.30.0
  - UI will be released in 4.31.0
- Add "End user authentication" and "macOS Setup Assistant" sections so the engineering team members at Fleet can link to these sections when implementing these stories (will be released in 4.31.0): #10689 and #10518

noahtalerman commented:
Hey @spokanemac I need some help with documenting "how to" instructions for signing a bootstrap package (Apple requires that they're signed): https://github.com/fleetdm/fleet/pull/11192/files#diff-53f79a9bcbecfe28f586509c8c8012fc032b247860f51dcbc0c0254d1644a0b2R39-R41

Can you help me write this?

Also, looking at these docs, is there any overlap with the bootstrap package article you're working on? Could we use some of your article in here? Or link to it?

spokanemac commented:
I need to test the signing steps in my [BYO config manager](https://docs.google.com/document/d/1-meXa-N3tZypVoKBVSdD1Dnc3GtSmSSBx5LiBi4dD7I/edit#heading=h.d9z6gsnf33s5), but I think we can use them.

noahtalerman commented:
> I need to test the signing steps in my [BYO config manager](https://docs.google.com/document/d/1-meXa-N3tZypVoKBVSdD1Dnc3GtSmSSBx5LiBi4dD7I/edit#heading=h.d9z6gsnf33s5), but I think we can use them.

@spokanemac oh nice! When we publish the article, I think those instructions should replace the instructions in the docs (link to article instead).

When do you think we'll publish the article?

zhumo commented on docs/Using-Fleet/MDM-macOS-setup.md:
```diff
@@ -0,0 +1,75 @@
+
+## End user authentication
+
+> This feature is currently in development.
```
What's the point of this and the other similar one below?

noahtalerman replied:
Features are currently being implemented. I want to give engineers working on the UI a valid URL to link to.

zhumo commented on docs/Using-Fleet/MDM-macOS-setup.md:
```diff
@@ -0,0 +1,75 @@
+
+In Fleet you can customize the first time macOS setup experience for your end users.
+
+You can require end users to authenticate with your identity provider (IdP) and agree to and end user license agreement (EULA) before they can use their new Mac.
```
I would recommend turning this into a bulleted list. Those are more scannable and signals to the reader to pay attention. One additional clue to use bulleted lists here is the repeated similar construction: you can... you can...

noahtalerman replied:
Great point. I turned this into a list

zhumo commented on docs/Using-Fleet/MDM-macOS-setup.md:
```diff
@@ -0,0 +1,75 @@
+
+You can customize the macOS Setup Assistant and choose to show or hide specific panes.
+
+Also, you can bootstrap new Macs with your configuration management tool of choice (ex. Munki, Chef, or Puppet).
```
I think we should be very literal. Strictly speaking, this is not the only thing we are enabling, right? We are allowing you to run an script at set up time. As part of this, you can install an agent, you can install a million other apps, or you can auto-popup a rickroll.

noahtalerman replied:
Hey @zhumo I updated this sentence to communicate that bootstrap package allows for "full control over the setup experience by installing tools like Puppet, Munki, DEP notify, custom scrips, and more."

What do you think?

zhumo commented on docs/Using-Fleet/MDM-macOS-setup.md:
```diff
@@ -0,0 +1,75 @@
+> This feature is currently in development.
+
+<meta name="pageOrderInSection" value="1504">
+<meta name="title" value="MDM commands">
```
Need to adjust the title here?

zhumo commented on docs/Using-Fleet/MDM-macOS-setup.md:
```diff
@@ -0,0 +1,75 @@
+
+_Available in Fleet Premium_
+
+Fleet supports installing a single package on macOS hosts that automatically enroll to Fleet.
```
I think we need to describe more about this feature. What does it do exactly? Why/when should I use it? Once you tell me that, then you tell me how to do it.

noahtalerman replied:
I updated this section to describe what the bootstrap package, why you should use it (to install tools like Puppet Munki and custom scripts), and added some examples of what organizations deploy.

What do you think?

noahtalerman commented:
Thanks for the review @zhumo!

I will ping you when this is ready for review.

Working on addressing this feedback:
-  https://github.com/fleetdm/fleet/pull/11192#discussion_r1169189633
- https://github.com/fleetdm/fleet/pull/11192#discussion_r1169191987

spokanemac commented:
@noahtalerman I hope my additions were applied correctly.

noahtalerman commented on docs/Using-Fleet/MDM-macOS-setup.md:
````diff
@@ -0,0 +1,120 @@
+     - Add the resulting `.p12` to your keychain.
+2. Sign your `pkg` with a valid Developer ID certificate:\
+\
+```productsign --sign "Developer ID Installer: Your Developer Name (SerialNumber)" /path/to/your.pkg /path/to/your-signed.pkg``` \
````
@spokanemac is there a UI option for signing? Just curious.

noahtalerman replied:
Gotcha. Nice. Yeah, I think starting with the productsign utility makes sense. Looks like it's include on Macs by default.

noahtalerman commented:
@spokanemac they did! Thanks.

I left some comments/feedback above.

spokanemac commented on docs/Using-Fleet/MDM-macOS-setup.md:
````diff
@@ -0,0 +1,120 @@
+\
+```productsign --sign "Developer ID Installer: Your Developer Name (SerialNumber)" /path/to/your.pkg /path/to/your-signed.pkg``` \
+
+3. Upload the package to Apple's notary service for validation: \
````
This might only apply to system extensions (formerly kernel extensions), but I was under the impression that all apps and pkgs need to be notarized but not stapled. [https://developer.apple.com/developer-id/](https://developer.apple.com/developer-id/)  I'll see if I can find more information to confirm.

noahtalerman replied:
Thanks! 

Yeah, I did a quick "installenterpriseapplication notarize" search in MacAdmins slack and found a couple messages that say you don't have to notarize. For example: https://macadmins.slack.com/archives/C19MR7EM9/p1633637758046800

I think we should test this. @roperzh, for when you're back, do you know if the bootstrap package needs to be notarized?

roperzh replied:
@noahtalerman It doesn't need to be notarized. Two things are key though, otherwise the installation fails:

1. It needs to be signed with a valid developer ID certificate
2. It needs to be a "distribution" style package

`2` depends on how you're building your package. Tools like Munki and [macadmins/installapplications](https://github.com/macadmins/installapplications) (`installapplications` uses Munki) allow you to set up that easily.

I can give you more details about how to do it if it's helpful.

noahtalerman replied:
@roperzh welcome back! 

Do the docs cover the two key things you outlined above? https://fleetdm.com/docs/using-fleet/mdm-macos-setup#bootstrap-package

roperzh replied:
thank you!

> Do the docs cover the two key things you outlined above? https://fleetdm.com/docs/using-fleet/mdm-macos-setup#bootstrap-package

I think it doesn't **explicitly** cover `2`. I haven't been able to find official documentation, but this is a good summary: https://scriptingosx.com/2017/09/on-distribution-packages/

In short, the package needs to be built in a "special" way.

noahtalerman commented:
Hey @spokanemac I updated the signing instructions: https://github.com/fleetdm/fleet/pull/11192/files#diff-53f79a9bcbecfe28f586509c8c8012fc032b247860f51dcbc0c0254d1644a0b2R49-R74

What do you think?

For now, I went with the assumption that we don't have to notarize the package. If we find that we're wrong after testing, we can update the docs.

I also modified the signing instructions so that they only include the Apple Developer account option. I think most Fleet customers will have an account.

spokanemac reviewed (approved):
This is a great distillation of my instructions. I might go into greater detail in a blog post.

noahtalerman commented:
Hey @zhumo! I merged in these docs so that we can use them during an MDM beta call today.

Heads up, I addressed your following feedback but you haven't had a chance to rereview:
-  https://github.com/fleetdm/fleet/pull/11192#discussion_r1169189633
- https://github.com/fleetdm/fleet/pull/11192#discussion_r1169191987

Please leave any suggestions/requested changes in this PR and I'll open a new PR to address them.
